app/controllers/discourse_subscriptions/admin/plans_controller.rb

```diff
@@ -70,16 +77,39 @@ def show
       def update
         begin
           plan =
-            ::Stripe::Price.update(
-              params[:id],
-              nickname: params[:nickname],
-              active: params[:active],
-              metadata: {
-                group_name: params[:metadata][:group_name],
-                trial_period_days: params[:trial_period_days],
-              },
-            )
-
+            ::Stripe::Price.update(params[:id], metadata: { group_name: params[:buffered_group] })
+          group = ::Group.find_by_name(params[:buffered_group])
+
+          unless group.nil?
+            if plan.recurring.nil?
+              checkout_sessions = get_checkout_sessions
+              checkout_sessions.each do |cs|
+                if cs.status == "complete"
+                  customer = Customer.find_by(customer_id: cs[:id])
+                  unless customer.nil?
+                    line_items =
+                      ::Stripe::Checkout::Session.list_line_items(cs[:id], { limit: 100 })
+                    line_items.each do |item|
+                      group = plan_group(item[:price]) unless item[:price].nil?
+                      group.add(customer.user) unless group.nil?
+                    end
+                  end
+                end
+              end
+              render_json_dump plan
+              return
+            end
+            subscriptions = get_subscription_data(params[:id])
+            discourse_customers = Customer.left_outer_joins(:user, :invite).all
+
+            subscriptions.each do |sub|
+              discourse_customers.each do |customer|
+                if sub[:customer][:id] == customer.customer_id && customer.user
+                  group.add(customer.user)
+                end
+              end
+            end
```

Comment:
In this section you can use eager loading to prevent N+1 queries.
Also you're doing a nested loop, and this is less than ideal for performance.

This section can be simplified and optimized by using batches as well:

```suggestion
customer_ids = get_subscription_data.map { |sub| sub[:customer][:id] }
customers = Customer.includes(:user, :invite).where(customer_id: customer_ids).where.not(user: nil)
customers.find_each(batch_size: 100)  { |c| group.add(c.user) }
```

Reply:
Overall I would focus on reducing nesting as a first step. Second step would be moving pieces of code into smaller functions so the method becomes easier to read.

Here are a couple of ideas.
Reducing nesting:
https://gist.github.com/jmperez127/348276777e000954183d402a90e32db9

Extra methods:
https://gist.github.com/jmperez127/34c3b9d8dbc674bf462f44406773df7e

I haven't tested this code, but it should give you an idea of what to improve.